dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -26,97 +30,278 @@ struct AsyncTasks
+                // We can hold the cancel handle here to prevent it from destructing, but it is not necessary.
+                (*p)();
+                // We don't erase from `tasks` here, since we won't capture `this`
+            },
+            0,
+            0);
         if (res)
         {
-            std::scoped_lock l(mtx);
-            futures[k] = p->get_future();
-            start_time[k] = getCurrentMillis();
+            tasks.insert({k, Elem(p->get_future(), getCurrentMillis(), std::move(triggered))});
         }
         return res;
     }
 
-    bool isScheduled(Key key) const
-    {
-        std::scoped_lock l(mtx);
-        return futures.contains(key);
-    }
-
-    bool isReady(Key key) const
+    TaskState queryState(Key key) const
     {
         using namespace std::chrono_literals;
         std::scoped_lock l(mtx);
-        auto it = futures.find(key);
-        if (it == futures.end())
-            return false;
-        return it->second.wait_for(0ms) == std::future_status::ready;
+        auto it = tasks.find(key);
+        if (it == tasks.end())
+            return TaskState::NotScheduled;
+        if (!it->second.triggered->load())
+            return TaskState::InQueue;
+        if (it->second.fut.wait_for(0ms) == std::future_status::ready)
+            return TaskState::Finished;
+        return TaskState::Running;
     }
 
-    R fetchResult(Key key)
-    {
-        std::unique_lock<std::mutex> l(mtx);
-        auto it = futures.find(key);
-        RUNTIME_CHECK_MSG(it != futures.end(), "fetchResult meets empty key");
-        auto fut = std::move(it->second);
-        futures.erase(it);
-        start_time.erase(key);
-        l.unlock();
-        return fut.get();
-    }
+    bool isScheduled(Key key) const { return queryState(key) != TaskState::NotScheduled; }
+
+    bool isInQueue(Key key) const { return queryState(key) == TaskState::InQueue; }
+
+    bool isRunning(Key key) const { return queryState(key) == TaskState::Running; }
+
+    bool isReady(Key key) const { return queryState(key) == TaskState::Finished; }
 
     uint64_t queryElapsed(Key key)
     {
         std::scoped_lock<std::mutex> l(mtx);
-        auto it2 = start_time.find(key);
-        RUNTIME_CHECK_MSG(it2 != start_time.end(), "queryElapsed meets empty key");
-        return getCurrentMillis() - it2->second;
+        auto it = tasks.find(key);
+        RUNTIME_CHECK_MSG(it != tasks.end(), "queryElapsed meets empty key");
+        return getCurrentMillis() - it->second.start_ts;
     }
 
-    uint64_t queryStartTime(Key key)
+    uint64_t queryStartTime(Key key, bool throw_if_not_found = true)
     {
         std::scoped_lock<std::mutex> l(mtx);
-        auto it2 = start_time.find(key);
-        RUNTIME_CHECK_MSG(it2 != start_time.end(), "queryElapsed meets empty key");
-        return it2->second;
+        auto it = tasks.find(key);
+        if unlikely (it == tasks.end())
+        {
+            if (throw_if_not_found)
```

Comment:
I prefer throw here because we don't RUNTIME_CHECK throw_if_not_found but `unlikely (it == tasks.end())`. However, we first check unlikely (it == tasks.end())` because it is more unlikely.